relay-server/src/processing/transactions/mod.rs

```diff
@@ -0,0 +1,592 @@
+        let mut extracted_metrics = ProcessingExtractedMetrics::new();
+
+        let mut work = work.try_map(|w, _| process::expand(w))?;
+
+        let mut run_dynamic_sampling = false;
+        work.try_modify(|work, record_keeper| {
+
+            let profile_id = profile::filter(work, record_keeper, &ctx, project_id);
+
+            let event = &mut work.transaction.0;
+            profile::transfer_id(event, profile_id);
+            profile::remove_context_if_rate_limited(event, scoping, ctx);
+
+            utils::dsc::validate_and_set_dsc(&mut work.headers, event, &mut ctx);
+
+            utils::event::finalize(
+                &work.headers,
+                event,
+                work.attachments.iter(),
+                &mut metrics,
+                ctx.config,
+            )?;
+
+            work.flags.fully_normalized = utils::event::normalize(
+                &work.headers,
+                event,
+                EventFullyNormalized(work.flags.fully_normalized),
+                project_id,
+                &ctx,
+                &self.geoip_lookup,
+            )?.0;
+
+            let filter_run = utils::event::filter(&work.headers, event, &ctx)
+                .map_err(ProcessingError::EventFiltered)?;
+
+            // Always run dynamic sampling on processing Relays,
+            // but delay decision until inbound filters have been fully processed.
+            // Also, we require transaction metrics to be enabled before sampling.
+            run_dynamic_sampling = (matches!(filter_run, FiltersStatus::Ok)
+                || ctx.config.processing_enabled())
+                && matches!(&ctx.project_info.config.transaction_metrics, Some(ErrorBoundary::Ok(c)) if c.is_enabled());
+
+            Ok::<_,Error>(())
+        })?;
+
+        let sampling_result = match run_dynamic_sampling {
+            true => {
+                #[allow(unused_mut)]
+                let mut reservoir = ReservoirEvaluator::new(Arc::clone(&self.reservoir_counters));
+                #[cfg(feature = "processing")]
+                if let Some(quotas_client) = self.quotas_client.as_ref() {
+                    reservoir.set_redis(work.scoping().organization_id, quotas_client);
+                }
+                utils::dynamic_sampling::run(
+                    work.headers.dsc(),
+                    &work.transaction.0,
+                    &ctx,
+                    Some(&reservoir),
+                )
+                .await
+            }
+            false => SamplingResult::Pending,
+        };
+
+        relay_statsd::metric!(
+            counter(RelayCounters::SamplingDecision) += 1,
+            decision = sampling_result.decision().as_str(),
+            item = "transaction"
+        );
+
+        #[cfg(feature = "processing")]
+        let server_sample_rate = sampling_result.sample_rate();
+
+        if let Some(outcome) = sampling_result.into_dropped_outcome() {
+            return process::drop_after_sampling(
```

Comment:
Similar to my comment for the function, I'd attempt to split this into multiple part and show the individual things which are happening in this function here, so a reader on first glance can guess better what is happening:

Pseudo:
```rs
let (remaining, metrics) = dynamic_sampling::drop_not_sampled(work);
self.limiter.enforce_limits(remaining)?;
return Ok(Output { main: remaining, metrics })
```

Reply:
Please take another look and let me know what you think.